worker/remoterelations/remoteapplicationworker.go

```diff
@@ -149,53 +149,76 @@ func (w *remoteApplicationWorker) loop() (err error) {
 	if !ok {
 		return nil
 	}
+	// On the consuming side, inform the remote side the relation is dying
+	// (but only if we are killing the relation due to it dying, not because
+	// it is suspended).
+	if !w.isConsumerProxy {
+		change := params.RemoteRelationChangeEvent{
+			RelationToken:    relation.relationToken,
+			Life:             params.Dying,
+			ApplicationToken: relation.applicationToken,
+			Macaroons:        macaroon.Slice{relation.macaroon},
+		}
+		if err := w.remoteModelFacade.PublishRelationChange(change); err != nil {
+			return errors.Annotatef(err, "publishing relation dying %+v to remote model %v", change, w.remoteModelUUID)
+		}
+	}
+	return nil
+}
+
+func (w *remoteApplicationWorker) processRelationSuspended(key string, relations map[string]*relation) error {
+	logger.Debugf("relation %v suspended", key)
+	relation, ok := relations[key]
+	if !ok {
+		return nil
+	}
+
 	// For suspended relations on the consuming side
 	// we want to keep the remote lifecycle watcher
 	// so we know when the relation is resumed.
-	// If we really are dying and not just suspended, then
-	// stop the lifecycle watcher and delete the relation.
-	if dying || w.isConsumerProxy {
-		// We are on the offering side so the relation can only be dying, or
-		// we are on the consuming side and are not suspended so must be dying.
-		delete(relations, key)
+	if w.isConsumerProxy {
 		if err := worker.Stop(relation.remoteRrw); err != nil {
 			logger.Warningf("stopping remote relations worker for %v: %v", key, err)
 		}
 		relation.remoteRuw = nil
+		delete(relations, key)
 	}
-	// For the unit watchers, check to see if these are nil before stopping.
-	// They will be nil if the relation was suspended and then we kill it for real.
+
 	if relation.localRuw != nil {
 		if err := worker.Stop(relation.localRuw); err != nil {
 			logger.Warningf("stopping local relation unit worker for %v: %v", key, err)
 		}
 		relation.localRuw = nil
 	}
+	return nil
+}
 
-	if !dying {
-		// If only suspended, no need to publish a dying event to the other side.
+func (w *remoteApplicationWorker) processRelationRemoved(key string, relations map[string]*relation) error {
+	logger.Debugf("relation %v removed", key)
+	relation, ok := relations[key]
+	if !ok {
 		return nil
 	}
 
-	// On the consuming side, inform the remote side the relation is dying
-	// (but only if we are killing the relation due to it dying, not because
-	// it is suspended).
-	if !w.isConsumerProxy && dying {
-		change := params.RemoteRelationChangeEvent{
-			RelationToken:    relation.relationToken,
-			Life:             params.Dying,
-			ApplicationToken: relation.applicationToken,
-			Macaroons:        macaroon.Slice{relation.macaroon},
-		}
-		if err := w.remoteModelFacade.PublishRelationChange(change); err != nil {
-			return errors.Annotatef(err, "publishing relation departed %+v to remote model %v", change, w.remoteModelUUID)
+	if err := worker.Stop(relation.remoteRrw); err != nil {
```

Comment:
If we check localRuw is nil below, we should check is remoteRuw is nil here

Reply:
that's unnecessary, relations[key] would be empty if remoteRuw were nil (also this is ~copy-and-paste from the old method)